File: src/core/TrustEnvironment.ts
/**
 * KnectIQ SelectiveTRUST® - Trust Environment
 *
 * A Trust Environment manages a collection of trust relationships
 * between provisioned devices. Part of the DASB control plane.
 *
 * Key Principles:
 * - Manages trust relationships, not data
 * - Real-time validation
 * - No access to encrypted data pathways
 * - Device-level trust without centralized key storage
 */

import crypto from 'crypto';

export interface TrustEnvironmentConfig {
  name: string;
  maxDevices: number;
  sessionTimeout: number;
  requiredTrustScore: number;
  allowedOperations: Set<string>;
}

export interface DeviceIdentity {
  deviceId: string;
  publicIdentity: Buffer;
  capabilities: string[];
  registered: number;
  lastSeen: number;
  trustScore: number;
}

export interface RelationshipMetadata {
  relationshipId: string;
  deviceA: string;
  deviceB: string;
  established: number;
  lastValidated: number;
  active: boolean;
}

/**
 * Trust Environment
 *
 * Manages collections of trust relationships between devices.
 * Does NOT handle encryption or data - only trust management.
 */
export class TrustEnvironment {
  private environmentId: string;
  private config: TrustEnvironmentConfig;
  private mode: 'fips-140-2' | 'standard';
  private devices: Map<string, DeviceIdentity> = new Map();
  private relationships: Map<string, RelationshipMetadata> = new Map();
  private validationCache: Map<string, { valid: boolean; timestamp: number }> = new Map();

  constructor(
    environmentId: string,
    config: TrustEnvironmentConfig,
    mode: 'fips-140-2' | 'standard' = 'standard'
  ) {
    this.environmentId = environmentId;
    this.config = config;
    this.mode = mode;
  }

  /**
   * Register a device in this trust environment
   */
  public registerDevice(
    deviceId: string,
    publicIdentity: Buffer,
    capabilities: string[] = []
  ): void {
    if (this.devices.size >= this.config.maxDevices) {
      throw new Error('Trust environment has reached maximum device capacity');
    }

    if (this.devices.has(deviceId)) {
      throw new Error(`Device ${deviceId} already registered in this environment`);
    }

    const identity: DeviceIdentity = {
      deviceId,
      publicIdentity,
      capabilities,
      registered: Date.now(),
      lastSeen: Date.now(),
      trustScore: 100 // Initial trust score
    };

    this.devices.set(deviceId, identity);
  }

  /**
   * Unregister a device from this trust environment
   */
  public unregisterDevice(deviceId: string): void {
    // Remove all relationships involving this device
    for (const [relationshipId, relationship] of this.relationships.entries()) {
      if (relationship.deviceA === deviceId || relationship.deviceB === deviceId) {
        this.revokeRelationship(relationshipId);
      }
    }

    this.devices.delete(deviceId);
  }

  /**
   * Check if device is registered in this environment
   */
  public hasDevice(deviceId: string): boolean {
    return this.devices.has(deviceId);
  }

  /**
   * Get device identity
   */
  public getDevice(deviceId: string): DeviceIdentity | undefined {
    return this.devices.get(deviceId);
  }

  /**
   * Establish a relationship between two devices
   */
  public establishRelationship(
    deviceIdA: string,
    deviceIdB: string,
    relationshipId: string
  ): void {
    if (!this.devices.has(deviceIdA) || !this.devices.has(deviceIdB)) {
      throw new Error('Both devices must be registered in this environment');
    }

    if (deviceIdA === deviceIdB) {
      throw new Error('Cannot establish relationship with self');
    }

    const metadata: RelationshipMetadata = {
      relationshipId,
      deviceA: deviceIdA,
      deviceB: deviceIdB,
      established: Date.now(),
      lastValidated: Date.now(),
      active: true
    };

    this.relationships.set(relationshipId, metadata);
  }

  /**
   * Validate a relationship in real-time
   *
   * Per SelectiveTRUST® patent: "real-time validation for every
   * communication/transaction"
   */
  public validateRelationship(relationshipId: string): boolean {
    const relationship = this.relationships.get(relationshipId);

    if (!relationship || !relationship.active) {
      return false;
    }

    // Check if devices still exist
    const deviceA = this.devices.get(relationship.deviceA);
    const deviceB = this.devices.get(relationship.deviceB);

    if (!deviceA || !deviceB) {
      relationship.active = false;
      return false;
    }

    // Check trust scores meet requirements
    if (
      deviceA.trustScore < this.config.requiredTrustScore ||
      deviceB.trustScore < this.config.requiredTrustScore
    ) {
      return false;
    }

    // Update last validated timestamp
    relationship.lastValidated = Date.now();

    // Update device last seen
    deviceA.lastSeen = Date.now();
    deviceB.lastSeen = Date.now();

    return true;
  }

  /**
   * Revoke a relationship
   */
  public revokeRelationship(relationshipId: string): void {
    const relationship = this.relationships.get(relationshipId);

    if (relationship) {
      relationship.active = false;
    }

    this.relationships.delete(relationshipId);
    this.validationCache.delete(relationshipId);
  }

  /**
   * Update device trust score
   *
   * Trust scores can be updated based on device behavior,
   * security posture, compliance status, etc.
   */
  public updateDeviceTrustScore(deviceId: string, newScore: number): void {
    const device = this.devices.get(deviceId);

    if (device) {
      device.trustScore = Math.max(0, Math.min(100, newScore));

      // If trust score falls below threshold, invalidate relationships
      if (device.trustScore < this.config.requiredTrustScore) {
        for (const [relationshipId, relationship] of this.relationships.entries()) {
          if (relationship.deviceA === deviceId || relationship.deviceB === deviceId) {
            relationship.active = false;
          }
        }
      }
    }
  }

  /**
   * Get all relationships for a device
   */
  public getDeviceRelationships(deviceId: string): RelationshipMetadata[] {
    const relationships: RelationshipMetadata[] = [];

    for (const relationship of this.relationships.values()) {
      if (
        (relationship.deviceA === deviceId || relationship.deviceB === deviceId) &&
        relationship.active
      ) {
        relationships.push(relationship);
      }
    }

    return relationships;
  }

  /**
   * Validate operation is allowed under trust policy
   */
  public validateOperation(operation: string): boolean {
    return this.config.allowedOperations.has(operation);
  }

  /**
   * Get environment statistics
   */
  public getStatistics(): {
    environmentId: string;
    devices: number;
    relationships: number;
    mode: string;
  } {
    return {
      environmentId: this.environmentId,
      devices: this.devices.size,
      relationships: this.relationships.size,
      mode: this.mode
    };
  }

  /**
   * Get environment configuration
   */
  public getConfig(): TrustEnvironmentConfig {
    return { ...this.config };
  }

  /**
   * Cleanup expired validation cache
   */
  private cleanupCache(): void {
    const now = Date.now();
    const cacheTimeout = 60000; // 1 minute

    for (const [key, value] of this.validationCache.entries()) {
      if (now - value.timestamp > cacheTimeout) {
        this.validationCache.delete(key);
      }
    }
  }

  /**
   * Shutdown trust environment
   */
  public shutdown(): void {
    // Revoke all relationships
    for (const relationshipId of this.relationships.keys()) {
      this.revokeRelationship(relationshipId);
    }

    // Clear devices
    this.devices.clear();
    this.validationCache.clear();
  }
}

File: ios-app/SecureMessenger/SecureMessenger/SelectiveTRUST/TrustEnvironment.swift
//
//  TrustEnvironment.swift
//  SecureMessenger - KnectIQ SelectiveTRUST® Integration
//
//  Manages collections of trust relationships between provisioned devices
//  Patent-based: Device-level trust without PKI dependency
//

import Foundation

public class TrustEnvironment {

    // MARK: - Properties

    private let environmentId: String
    private var config: Config
    private let mode: DeviceSDK.OperatingMode

    private var devices: [String: DeviceIdentity] = [:]
    private var relationships: [String: RelationshipMetadata] = [:]

    private let queue = DispatchQueue(label: "com.securemessenger.trustenv")

    // MARK: - Types

    public struct Config {
        let name: String
        let maxDevices: Int
        let sessionTimeout: TimeInterval
        let requiredTrustScore: Int
        var allowedOperations: Set<String>

        public init(name: String, maxDevices: Int = 100, sessionTimeout: TimeInterval = 3600,
                    requiredTrustScore: Int = 80, allowedOperations: Set<String> = ["send", "receive", "read"]) {
            self.name = name
            self.maxDevices = maxDevices
            self.sessionTimeout = sessionTimeout
            self.requiredTrustScore = requiredTrustScore
            self.allowedOperations = allowedOperations
        }
    }

    public struct DeviceIdentity {
        let deviceId: String
        let publicIdentity: Data
        let capabilities: [String]
        let registered: Date
        var lastSeen: Date
        var trustScore: Int
    }

    public struct RelationshipMetadata {
        let relationshipId: String
        let deviceA: String
        let deviceB: String
        let established: Date
        var lastValidated: Date
        var active: Bool
    }

    // MARK: - Initialization

    public init(environmentId: String, config: Config, mode: DeviceSDK.OperatingMode = .standard) {
        self.environmentId = environmentId
        self.config = config
        self.mode = mode
    }

    // MARK: - Device Registration

    public func registerDevice(deviceId: String, publicIdentity: Data, capabilities: [String] = []) throws {
        try queue.sync {
            guard devices.count < config.maxDevices else {
                throw TrustEnvironmentError.maxDevicesReached
            }

            guard devices[deviceId] == nil else {
                throw TrustEnvironmentError.deviceAlreadyRegistered
            }

            let identity = DeviceIdentity(
                deviceId: deviceId,
                publicIdentity: publicIdentity,
                capabilities: capabilities,
                registered: Date(),
                lastSeen: Date(),
                trustScore: 100  // Initial perfect trust score
            )

            devices[deviceId] = identity
        }
    }

    public func unregisterDevice(deviceId: String) {
        queue.sync {
            // Remove all relationships involving this device
            let deviceRelationships = relationships.filter {
                $0.value.deviceA == deviceId || $0.value.deviceB == deviceId
            }

            for (relationshipId, _) in deviceRelationships {
                relationships.removeValue(forKey: relationshipId)
            }

            devices.removeValue(forKey: deviceId)
        }
    }

    public func hasDevice(deviceId: String) -> Bool {
        queue.sync {
            devices[deviceId] != nil
        }
    }

    public func getDevice(deviceId: String) -> DeviceIdentity? {
        queue.sync {
            devices[deviceId]
        }
    }

    // MARK: - Relationship Management

    public func establishRelationship(deviceIdA: String, deviceIdB: String, relationshipId: String) throws {
        try queue.sync {
            guard devices[deviceIdA] != nil && devices[deviceIdB] != nil else {
                throw TrustEnvironmentError.deviceNotFound
            }

            guard deviceIdA != deviceIdB else {
                throw TrustEnvironmentError.cannotEstablishSelfRelationship
            }

            let metadata = RelationshipMetadata(
                relationshipId: relationshipId,
                deviceA: deviceIdA,
                deviceB: deviceIdB,
                established: Date(),
                lastValidated: Date(),
                active: true
            )

            relationships[relationshipId] = metadata
        }
    }

    /// Real-time trust validation
    /// Patent claim: "Real-time validation for every communication/transaction"
    public func validateRelationship(relationshipId: String) -> Bool {
        queue.sync {
            guard var relationship = relationships[relationshipId],
                  relationship.active else {
                return false
            }

            // Check if devices still exist
            guard let deviceA = devices[relationship.deviceA],
                  let deviceB = devices[relationship.deviceB] else {
                relationship.active = false
                relationships[relationshipId] = relationship
                return false
            }

            // Check trust scores meet requirements
            guard deviceA.trustScore >= config.requiredTrustScore &&
                  deviceB.trustScore >= config.requiredTrustScore else {
                return false
            }

            // Update validation timestamp
            relationship.lastValidated = Date()
            relationships[relationshipId] = relationship

            // Update device last seen
            devices[relationship.deviceA]?.lastSeen = Date()
            devices[relationship.deviceB]?.lastSeen = Date()

            return true
        }
    }

    public func revokeRelationship(relationshipId: String) {
        queue.sync {
            if var relationship = relationships[relationshipId] {
                relationship.active = false
                relationships[relationshipId] = relationship
            }
            relationships.removeValue(forKey: relationshipId)
        }
    }

    public func getDeviceRelationships(deviceId: String) -> [RelationshipMetadata] {
        queue.sync {
            relationships.values.filter {
                ($0.deviceA == deviceId || $0.deviceB == deviceId) && $0.active
            }
        }
    }

    // MARK: - Trust Score Management

    public func updateDeviceTrustScore(deviceId: String, newScore: Int) {
        queue.sync {
            guard var device = devices[deviceId] else { return }

            device.trustScore = max(0, min(100, newScore))
            devices[deviceId] = device

            // If trust score falls below threshold, deactivate relationships
            if device.trustScore < config.requiredTrustScore {
                for (relationshipId, var relationship) in relationships {
                    if relationship.deviceA == deviceId || relationship.deviceB == deviceId {
                        relationship.active = false
                        relationships[relationshipId] = relationship
                    }
                }
            }
        }
    }

    // MARK: - Operations

    public func validateOperation(operation: String) -> Bool {
        config.allowedOperations.contains(operation)
    }

    // MARK: - Statistics

    public func getStatistics() -> Statistics {
        queue.sync {
            Statistics(
                environmentId: environmentId,
                devices: devices.count,
                relationships: relationships.count,
                mode: mode
            )
        }
    }

    public struct Statistics {
        public let environmentId: String
        public let devices: Int
        public let relationships: Int
        public let mode: DeviceSDK.OperatingMode
    }

    public func getConfig() -> Config {
        config
    }

    // MARK: - Shutdown

    public func shutdown() {
        queue.sync {
            // Revoke all relationships
            for relationshipId in relationships.keys {
                revokeRelationship(relationshipId: relationshipId)
            }

            devices.removeAll()
        }
    }

    // MARK: - Errors

    public enum TrustEnvironmentError: Error {
        case maxDevicesReached
        case deviceAlreadyRegistered
        case deviceNotFound
        case cannotEstablishSelfRelationship
    }
}
